src/bsk_rl/obs/observations.py

```diff
@@ -0,0 +1,456 @@
+
+
+class ObservationBuilder:
+
+    def __init__(self, satellite: "Satellite", obs_type: type = np.ndarray) -> None:
+        """Satellite subclass for composing observations.
+
+        Args:
+            satellite: Satellite to observe
+            obs_type: Datatype of satellite's returned observation. Can be ``np.ndarray``
+                (default), ``dict``, or ``list``.
+        """
+        self.obs_type = obs_type
+        self.obs_dict_cache = None
+        self.obs_cache_time = 0.0
+        self.satellite = satellite
+        self.simulator: "Simulator"
+        self.observation_spec = deepcopy(self.satellite.observation_spec)
+        name_counts = {}
+        for obs in self.observation_spec:
+            if obs.name in name_counts:
+                name_counts[obs.name] += 1
+                obs.name += f"_{name_counts[obs.name]}"
+            else:
+                name_counts[obs.name] = 1
+            obs.link_satellite(self.satellite)
+
+    def reset_post_sim(self) -> None:
+        """Perform any once-per-episode setup."""
+        self.simulator = self.satellite.simulator  # already a proxy
+        self.obs_dict_cache = None
+        for obs in self.observation_spec:
+            obs.link_simulator(self.simulator)  # already a proxy
+            obs.reset_post_sim()
+
+    def obs_dict(self) -> dict[str, Any]:
+        """Human-readable observation format.
+
+        Cached so only computed once per timestep.
+        """
+        if (
+            self.obs_dict_cache is None
+            or self.simulator.sim_time != self.obs_cache_time
+        ):
+            self.obs_dict_cache = {
+                obs.name: obs.get_obs() for obs in self.observation_spec
+            }
+            self.obs_cache_time = self.simulator.sim_time
+        return deepcopy(self.obs_dict_cache)
+
+    def obs_ndarray(self) -> np.ndarray:
+        """Numpy vector observation format."""
+        _, obs = vectorize_nested_dict(self.obs_dict())
+        return obs
+
+    def obs_array_keys(self) -> list[str]:
+        """Get the keys of the obs_ndarray."""
+        keys, _ = vectorize_nested_dict(self.obs_dict())
+        return keys
+
+    def obs_list(self) -> list:
+        """List observation format."""
+        return list(self.obs_ndarray())
+
+    def get_obs(self) -> Union[dict, np.ndarray, list]:
+        """Update the observation."""
+        if self.obs_type is dict:
+            return self.obs_dict()
+        elif self.obs_type is np.ndarray:
+            return self.obs_ndarray()
+        elif self.obs_type is list:
+            return self.obs_list()
+        else:
+            raise ValueError(f"Invalid observation type: {self.obs_type}")
+
+    @property
+    def observation_space(self) -> spaces.Space:
+        """Space of the observation."""
+        obs = self.get_obs()
+        if isinstance(obs, (list, np.ndarray)):
+            return spaces.Box(low=-1e16, high=1e16, shape=obs.shape, dtype=np.float64)
```

Comment:
Can you call `obs_dict_to_space(obs)` here as you did in line 127 considering that `obs_dict_to_space` check for types?

Reply:
good observation (pun intended)